From c9005a5013c3880f53fc34559d05f7691702ffed Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Mon, 13 Apr 2015 15:58:18 -0500
Subject: [PATCH] Force load chunks for specific entities that fly through


diff --git a/src/main/java/net/minecraft/server/EntityEnderPearl.java b/src/main/java/net/minecraft/server/EntityEnderPearl.java
index 9376a1d..a29f65f 100644
--- a/src/main/java/net/minecraft/server/EntityEnderPearl.java
+++ b/src/main/java/net/minecraft/server/EntityEnderPearl.java
@@ -63,4 +63,24 @@ public class EntityEnderPearl extends EntityProjectile {
             this.die();
         }
     }
+
+    // PaperSpigot start - Force load chunks for specific entities that fly through
+    public void h() {
+        if (world.paperSpigotConfig.loadUnloadedEnderPearls) {
+            ChunkProviderServer chunkProvider = (ChunkProviderServer) world.chunkProvider;
+            boolean before = chunkProvider.forceChunkLoad;
+            chunkProvider.forceChunkLoad = true;
+
+            for (int cx = (int) locX >> 4; cx <= (int) (locX + motX) >> 4; ++cx) {
+                for (int cz = (int) locZ >> 4; cz <= (int) (locZ + motZ) >> 4; ++cz) {
+                    world.getChunkAt(cx, cz);
+                }
+            }
+
+            chunkProvider.forceChunkLoad = before;
+        }
+
+        super.h();
+    }
+    // PaperSpigot end
 }
diff --git a/src/main/java/net/minecraft/server/EntityFallingBlock.java b/src/main/java/net/minecraft/server/EntityFallingBlock.java
index 51d2505..ce89938 100644
--- a/src/main/java/net/minecraft/server/EntityFallingBlock.java
+++ b/src/main/java/net/minecraft/server/EntityFallingBlock.java
@@ -72,6 +72,21 @@ public class EntityFallingBlock extends Entity {
         if (this.id.getMaterial() == Material.AIR) {
             this.die();
         } else {
+            // PaperSpigot start - Force load chunks for specific entities that fly through
+            if (world.paperSpigotConfig.loadUnloadedFallingBlocks) {
+                ChunkProviderServer chunkProvider = (ChunkProviderServer) world.chunkProvider;
+                boolean before = chunkProvider.forceChunkLoad;
+                chunkProvider.forceChunkLoad = true;
+
+                for (int cx = (int) locX >> 4; cx <= (int) (locX + motX) >> 4; ++cx) {
+                    for (int cz = (int) locZ >> 4; cz <= (int) (locZ + motZ) >> 4; ++cz) {
+                        world.getChunkAt(cx, cz);
+                    }
+                }
+
+                chunkProvider.forceChunkLoad = before;
+            }
+            // PaperSpigot end
             this.lastX = this.locX;
             this.lastY = this.locY;
             this.lastZ = this.locZ;
diff --git a/src/main/java/net/minecraft/server/EntityTNTPrimed.java b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
index 8aa1663..380fcca 100644
--- a/src/main/java/net/minecraft/server/EntityTNTPrimed.java
+++ b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
@@ -53,6 +53,21 @@ public class EntityTNTPrimed extends Entity {
 
     public void h() {
         if (world.spigotConfig.currentPrimedTnt++ > world.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
+        // PaperSpigot start - Force load chunks for specific entities that fly through
+        if (world.paperSpigotConfig.loadUnloadedTNTEntities) {
+            ChunkProviderServer chunkProvider = (ChunkProviderServer) world.chunkProvider;
+            boolean before = chunkProvider.forceChunkLoad;
+            chunkProvider.forceChunkLoad = true;
+
+            for (int cx = (int) locX >> 4; cx <= (int) (locX + motX) >> 4; ++cx) {
+                for (int cz = (int) locZ >> 4; cz <= (int) (locZ + motZ) >> 4; ++cz) {
+                    world.getChunkAt(cx, cz);
+                }
+            }
+
+            chunkProvider.forceChunkLoad = before;
+        }
+        // PaperSpigot end
         this.lastX = this.locX;
         this.lastY = this.locY;
         this.lastZ = this.locZ;
@@ -95,8 +110,21 @@ public class EntityTNTPrimed extends Entity {
         server.getPluginManager().callEvent(event);
 
         if (!event.isCancelled()) {
-            // give 'this' instead of (Entity) null so we know what causes the damage
-            this.world.createExplosion(this, this.locX, this.locY, this.locZ, event.getRadius(), event.getFire(), true);
+            // PaperSpigot start - Force load chunks for specific entities that fly through
+            if (world.paperSpigotConfig.loadUnloadedTNTEntities) {
+                ChunkProviderServer chunkProvider = (ChunkProviderServer) world.chunkProvider;
+                boolean before = chunkProvider.forceChunkLoad;
+                chunkProvider.forceChunkLoad = true;
+
+                // give 'this' instead of (Entity) null so we know what causes the damage
+                this.world.createExplosion(this, this.locX, this.locY, this.locZ, event.getRadius(), event.getFire(), true);
+
+                chunkProvider.forceChunkLoad = before;
+            } else {
+                // give 'this' instead of (Entity) null so we know what causes the damage
+                this.world.createExplosion(this, this.locX, this.locY, this.locZ, event.getRadius(), event.getFire(), true);
+            }
+            // PaperSpigot end
         }
         // CraftBukkit end
     }
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 2ad8614..dedc487 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1691,6 +1691,17 @@ public abstract class World implements IBlockAccess {
                 if (this.isChunkLoaded(k, i1)) {
                     entity.ag = true;
                     this.getChunkAt(k, i1).a(entity);
+                // PaperSpigot start - Force load chunks for specific entities that fly through
+                } else if ((entity instanceof EntityEnderPearl && entity.world.paperSpigotConfig.loadUnloadedEnderPearls) ||
+                        (entity instanceof EntityTNTPrimed && entity.world.paperSpigotConfig.loadUnloadedTNTEntities) ||
+                        (entity instanceof EntityFallingBlock && entity.world.paperSpigotConfig.loadUnloadedFallingBlocks)) {
+                    ChunkProviderServer chunkProvider = (ChunkProviderServer) this.chunkProvider;
+                    boolean before = chunkProvider.forceChunkLoad;
+                    chunkProvider.forceChunkLoad = true;
+                    entity.ag = true;
+                    this.getChunkAt(k, i1).a(entity);
+                    chunkProvider.forceChunkLoad = before;
+                // PaperSpigot end
                 } else {
                     entity.ag = false;
                 }
diff --git a/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java b/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java
index 4892310..f3d7129 100644
--- a/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java
+++ b/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java
@@ -235,10 +235,20 @@ public class PaperSpigotWorldConfig
         generateTemple = getBoolean( "generator-settings.temple", true );
         generateVillage = getBoolean( "generator-settings.village", true );
     }
-
+    
     public boolean optimizeDraining;
     private void optimizeDraining()
     {
         optimizeDraining = getBoolean( "optimize-draining", false );
     }
+
+    public boolean loadUnloadedEnderPearls;
+    public boolean loadUnloadedTNTEntities;
+    public boolean loadUnloadedFallingBlocks;
+    private void loadUnloaded()
+    {
+        loadUnloadedEnderPearls = getBoolean( "load-chunks.enderpearls", false );
+        loadUnloadedTNTEntities = getBoolean( "load-chunks.tnt-entities", false );
+        loadUnloadedFallingBlocks = getBoolean( "load-chunks.falling-blocks", false );
+    }
 }
diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index 6fea403..06145fc 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -12,6 +12,7 @@ import net.minecraft.server.EntityComplexPart;
 import net.minecraft.server.EntityCreature;
 import net.minecraft.server.EntityEnderCrystal;
 import net.minecraft.server.EntityEnderDragon;
+import net.minecraft.server.EntityEnderPearl; // PaperSpigot
 import net.minecraft.server.EntityFallingBlock; // PaperSpigot
 import net.minecraft.server.EntityFireball;
 import net.minecraft.server.EntityFireworks;
@@ -266,7 +267,9 @@ public class ActivationRange
 
         // PaperSpigot start - EAR backport
         // Never safe to skip fireworks or entities not yet added to chunk and we don't skip falling blocks
-        if ( !entity.isAddedToChunk() || entity instanceof EntityFireworks || entity instanceof EntityFallingBlock ) {
+        if ( !entity.isAddedToChunk() || entity instanceof EntityFireworks || entity instanceof EntityFallingBlock ||
+                (entity instanceof EntityEnderPearl && entity.world.paperSpigotConfig.loadUnloadedEnderPearls) ||
+                (entity instanceof EntityTNTPrimed && entity.world.paperSpigotConfig.loadUnloadedTNTEntities) ) {
             SpigotTimings.checkIfActiveTimer.stopTiming();
             return true;
         }
-- 
1.9.4.msysgit.2

