From c768ee87006c19c78bfafae7f802d866dc93673d Mon Sep 17 00:00:00 2001
From: Iceee <andrew@opticgaming.tv>
Date: Wed, 17 Jun 2015 00:57:31 -0700
Subject: [PATCH] Stop flowing lava and water when set to air


diff --git a/src/main/java/net/minecraft/server/BlockFlowing.java b/src/main/java/net/minecraft/server/BlockFlowing.java
index fd33a1e..2bb81ad 100644
--- a/src/main/java/net/minecraft/server/BlockFlowing.java
+++ b/src/main/java/net/minecraft/server/BlockFlowing.java
@@ -85,6 +85,7 @@ public class BlockFlowing extends BlockFluids {
                 l = j1;
                 if (j1 < 0) {
                     world.setAir(i, j, k);
+                    return; // PaperSpigot - Stop flowing lava and water when set to air
                 } else {
                     world.setData(i, j, k, j1, 2);
                     world.a(i, j, k, this, i1);
-- 
1.9.5.msysgit.1

