From 1f1421d2bc3913b3f5ee73cb73700098b926f7b5 Mon Sep 17 00:00:00 2001
From: DoctorDark <doctordark11@gmail.com>
Date: Sun, 4 Oct 2015 04:03:39 +0100
Subject: [PATCH] Add EquipmentSetEvent API


diff --git a/src/main/java/org/bukkit/event/inventory/EquipmentSetEvent.java b/src/main/java/org/bukkit/event/inventory/EquipmentSetEvent.java
new file mode 100644
index 0000000..18ab995
--- /dev/null
+++ b/src/main/java/org/bukkit/event/inventory/EquipmentSetEvent.java
@@ -0,0 +1,48 @@
+package org.bukkit.event.inventory;
+
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.bukkit.inventory.ItemStack;
+
+public class EquipmentSetEvent extends Event {
+
+    private static final HandlerList handlers = new HandlerList();
+
+    private final HumanEntity humanEntity;
+    private final int slot;
+    private final ItemStack previousItem;
+    private final ItemStack newItem;
+
+    public EquipmentSetEvent(HumanEntity humanEntity, int slot, ItemStack previousItem, ItemStack newItem) {
+        this.humanEntity = humanEntity;
+        this.slot = slot;
+        this.previousItem = previousItem;
+        this.newItem = newItem;
+    }
+
+    public HumanEntity getHumanEntity() {
+        return this.humanEntity;
+    }
+
+    public int getSlot() {
+        return this.slot;
+    }
+
+    public ItemStack getPreviousItem() {
+        return this.previousItem;
+    }
+
+    public ItemStack getNewItem() {
+        return this.newItem;
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
\ No newline at end of file
-- 
1.9.5.github.0

